File: DSA/Priyanshi/Day2_question2.java
import java.util.*;
public class Day2_question2 {
        public static void main(String[] args) {
            Scanner sc = new Scanner(System.in);
            System.out.println("Enter the number of leaderboard scores:");
            int n = sc.nextInt();
            int[] leaderboard = new int[n];
            System.out.println("Enter the leaderboard scores:");
            for (int i = 0; i < n; i++) {
                leaderboard[i] = sc.nextInt();
            }
            System.out.println("Enter the number of player scores:");
            int m = sc.nextInt();
            int[] player_scores = new int[m];
            System.out.println("Enter the player scores:");
            for (int i = 0; i < m; i++) {
                player_scores[i] = sc.nextInt();
            }
            int[] ranks = climbingLeaderboard(leaderboard, player_scores);
            System.out.println("Ranks:");
            for (int rank : ranks) {
                System.out.print(rank + " ");
            }
        }
    
        public static int[] climbingLeaderboard(int[] leaderboard, int[] player_scores) {
            int[] ranks = new int[player_scores.length];
            for (int i = 0; i < player_scores.length; i++) {
                int rank = 1;
                for (int j = 0; j < leaderboard.length; j++) {
                    if (player_scores[i] < leaderboard[j]) {
                        rank++;
                    }
                }
                ranks[i] = rank;
            }
            return ranks;
        }
    }

File: DSA/Arshdeep-Kaur/Day2_Question2.cpp
#include <iostream>
using namespace std;

void sortDescending(int arr[], int n) {
    for (int i = 0; i < n - 1; ++i) {
        for (int j = 0; j < n - i - 1; ++j) {
            if (arr[j] < arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

int removeDuplicates(int arr[], int n) {
    if (n == 0 || n == 1) return n;

    int temp[n];
    int j = 0;

    for (int i = 0; i < n - 1; ++i) {
        if (arr[i] != arr[i + 1]) {
            temp[j++] = arr[i];
        }
    }
    temp[j++] = arr[n - 1];

    for (int i = 0; i < j; ++i) {
        arr[i] = temp[i];
    }

    return j;
}

int* climbingLeaderboard(int leaderboard[], int lb_size, int player_scores[], int ps_size) {
    sortDescending(leaderboard, lb_size);
    int unique_count = removeDuplicates(leaderboard, lb_size);
    int* ranks = new int[ps_size];

    for (int i = 0; i < ps_size; ++i) {
        int score = player_scores[i];

        // Binary search for the correct position
        int low = 0, high = unique_count;

        while (low < high) {
            int mid = (low + high) / 2;
            if (leaderboard[mid] > score) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        ranks[i] = low + 1;
    }

    return ranks;
}

int main() {
    int leaderboard[100], lb_size, player_scores[100], ps_size;
    cout << "Enter the number of scores in the leaderboard: ";
    cin >> lb_size;
    cout << "Enter the scores in the leaderboard: ";
    for (int i = 0; i < lb_size; i++) {
        cin >> leaderboard[i];
    }

    cout << "Enter the number of scores of the player: ";
    cin >> ps_size;
    cout << "Enter the scores of the player: ";
    for (int i = 0; i < ps_size; i++) {
        cin >> player_scores[i];
    }

    int* ranks = climbingLeaderboard(leaderboard, lb_size, player_scores, ps_size);

    for (int i = 0; i < ps_size; ++i) {
        cout << ranks[i] << " ";
    }
    cout << endl;
}
